### PR TITLE
ZDocument: Import Document and related infrastructure from istoph/editor.

### PR DESCRIPTION
**Pull request by textshell**
This is a step along the way to a multi line text edit widget.

### CODE REVIEW

**Comment by textshell** [comment]
A rendered version of of the documentation is available at
https://tuiwidgets.namepad.de/wip/ZDocument.html

**Comment by textshell** [comment]
I'm also open to any kind of partial reviews. Like public interface or documentation, or any other part.

File: doc/ZDocument.rst [hunk 1]
```diff
@@ -0,0 +1,1186 @@
...
+--------------
+
+It is often useful for an application to store additional information associated with a line.
+This could for example be cached syntax highlighting information or generally additional information.
+
+To facilitate this, the document allows saving an additional object with each line, by using
+:cpp:func:`setLineUserData <void Tui::ZDocument::setLineUserData(int line, std::shared_ptr<Tui::ZDocumentLineUserData> userData)>`.
+
+Setting it or changes to it do not trigger change signals or increment the document or line revisions.
+The user is responsible for threading save usage of this object, when using snapshots with other
```
Comment (bziemons) [inline doc/ZDocument.rst:82]: -threading save
+thread safe

File: doc/ZDocument.rst [hunk 2]
```diff
@@ -0,0 +1,1186 @@
...
+threads than the document's thread.
+
+Currently the user-data is restored when using undo and redo, but this might change in the future.
```
Comment (bziemons) [inline doc/ZDocument.rst:85]: what does restored mean in this context? deleted and undeleted? Or instead: retained?

**Review by istoph — APPROVED** [review]

**Pull request merged**